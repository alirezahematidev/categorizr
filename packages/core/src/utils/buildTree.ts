import { FlatNode } from "$core/index";

type BuildTreeReturn<T extends FlatNode> = Omit<T, "parentId"> & {
  id: string;
  children: BuildTreeReturn<T>[];
};

export function buildTree<T extends FlatNode>(flatNodes: T[]): BuildTreeReturn<T>[] {
  const nodesMap = new Map<string, BuildTreeReturn<T>>();

  for (const flatNode of flatNodes) {
    const { id, parentId, ...rest } = flatNode;

    nodesMap.set(id, { id, children: [], ...rest } as BuildTreeReturn<T>);
  }

  for (const flatNode of flatNodes) {
    const { id, parentId } = flatNode;

    const node = nodesMap.get(id)!;

    if (parentId !== null) {
      const parent = nodesMap.get(parentId)!;
      parent.children!.push(node);
    }
  }

  const roots: BuildTreeReturn<T>[] = [];

  for (const node of nodesMap.values()) {
    if (!flatNodes.some((flatNode) => flatNode.id === node.id && flatNode.parentId !== null)) {
      roots.push(node);
    }
  }

  return roots;
}
