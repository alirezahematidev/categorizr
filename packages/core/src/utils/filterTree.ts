import { TreeNode } from "$core/index";

function filterTree<T extends TreeNode>(tree: readonly T[], predicate: (node: T) => boolean): T[] {
  const filteredNodes: T[] = [];

  function traverse(node: T | undefined) {
    if (!node) return;

    if (predicate(node)) filteredNodes.push(node);

    if (node.children) {
      for (const child of node.children) {
        traverse(child as T);
      }
    }
  }

  for (const node of tree) {
    traverse(node);
  }

  return filteredNodes;
}

export { filterTree };
