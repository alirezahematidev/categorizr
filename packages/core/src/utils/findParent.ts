import { TreeNode } from "$core/types";
import { noChildren } from "./helpers";

export function findParent<T extends TreeNode>(tree: T[], node: T): T | undefined {
  for (const branch of tree) {
    if (noChildren(branch)) continue;

    if (branch.children.some((child) => child.id === node.id)) return branch;

    const parent = findParent<T>(branch.children as T[], node);

    if (parent) return parent;
  }

  return undefined;
}
