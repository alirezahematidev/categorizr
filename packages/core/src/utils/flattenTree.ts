import type { TreeNode } from "$core/index";

type FlattenTreeReturn<T extends TreeNode> = Omit<T, "children"> & { parentId: string | null };

export function flattenTree<T extends TreeNode>(tree: readonly T[]): FlattenTreeReturn<T>[] {
  const flatNodes: FlattenTreeReturn<T>[] = [];

  function traverse(node: T | undefined, parentId: string | null) {
    if (!node) return;

    const { children, ...rest } = node;

    flatNodes.push({ ...rest, parentId });

    if (node.children) {
      for (const child of node.children) {
        traverse(child as T, node.id);
      }
    }
  }

  for (const node of tree) {
    traverse(node, null);
  }

  return flatNodes;
}
