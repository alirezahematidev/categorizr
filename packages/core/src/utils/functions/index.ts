export * from "./findNode";
export * from "./findParent";
export * from "./insertNode";
export * from "./safeInsertNode";
