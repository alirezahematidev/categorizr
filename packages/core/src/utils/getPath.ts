import { TreeNode } from "$core/index";

export function getPath<T extends TreeNode>(tree: readonly T[], id: string): string[] {
  const path: string[] = [];

  const findPath = (nodes: readonly T[], nodeId: string): boolean => {
    for (const node of nodes) {
      path.push(node.id);

      if (node.id === nodeId) return true;

      if (node.children && findPath(node.children as T[], nodeId)) return true;
    }

    return false;
  };

  findPath(tree, id);

  return path;
}
