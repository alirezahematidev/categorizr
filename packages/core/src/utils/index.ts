export * from "./buildTree";
export * from "./flattenTree";
export * from "./filterTree";
