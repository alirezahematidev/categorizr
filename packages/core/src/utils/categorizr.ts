import { DestId, InsertCallback, InsertCallbackWithError, MaybeNode, TreeNode } from "$core/types";
import { findParent, findNode, insertNode, safeInsertNode, removeNode } from "./functions";

export class Categorizr<T extends TreeNode> {
  private readonly tree: ReadonlyArray<T> = [];

  constructor(tree: T[]) {
    this.tree = tree;
  }

  public findParent(node: T): MaybeNode<T> {
    return findParent(this.tree, node);
  }

  public findNode<Id extends T["id"]>(id: Id): MaybeNode<T> {
    return findNode(this.tree, id);
  }

  public removeNode<Id extends T["id"]>(id: Id): T[] {
    return removeNode(this.tree, id);
  }

  public insertNode<const Id extends T["id"]>(destId: DestId<Id>, node: T, callback: InsertCallback<T>): void;
  public insertNode<const Id extends T["id"]>(destId: DestId<Id>, node: T): T[];
  public insertNode(destId: DestId<any>, node: any, callback?: InsertCallback<any>) {
    if (callback) return insertNode(this.tree, destId, node, callback);

    return insertNode(this.tree, destId, node);
  }

  public safeInsertNode<const Id extends T["id"]>(destId: DestId<Id>, node: T, callback: InsertCallbackWithError<T>): void;
  public safeInsertNode<const Id extends T["id"]>(destId: DestId<Id>, node: T): T[];
  public safeInsertNode(destId: DestId<any>, node: any, callback?: InsertCallbackWithError<any>) {
    if (callback) return safeInsertNode(this.tree, destId, node, callback);

    return safeInsertNode(this.tree, destId, node);
  }
}
