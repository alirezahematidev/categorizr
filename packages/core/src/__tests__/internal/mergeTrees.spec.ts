import { TreeNode } from "$core/index";
import { mergeTrees } from "$core/utils";
import { afterEach, describe, expect, it, vi } from "vitest";

describe("mergeTrees", async () => {
  afterEach(() => {
    vi.resetAllMocks();
  });

  it("should returns the merged trees", () => {
    const fn = vi.fn(mergeTrees);

    const tree1: TreeNode[] = [
      {
        id: "1",
        children: [
          { id: "2", children: [] },
          { id: "3", children: [] },
        ],
      },
      { id: "4", children: [{ id: "5", children: [] }] },
    ];

    const tree2: TreeNode[] = [
      { id: "1", children: [{ id: "6", children: [] }] },
      { id: "7", children: [{ id: "8", children: [] }] },
    ];

    const mergedTrees = fn(tree1, tree2);

    expect(mergedTrees).toStrictEqual([
      {
        id: "1",
        children: [
          { id: "2", children: [] },
          { id: "3", children: [] },
          { id: "6", children: [] },
        ],
      },
      { id: "4", children: [{ id: "5", children: [] }] },
      { id: "7", children: [{ id: "8", children: [] }] },
    ]);

    expect(mergedTrees).toMatchSnapshot();
  });

  it("should returns the merged complex trees", () => {
    const fn = vi.fn(mergeTrees);

    const tree1: TreeNode[] = [
      {
        id: "1",
        children: [
          {
            id: "2",
            children: [
              { id: "3", children: [] },
              { id: "4", children: [] },
            ],
          },
          { id: "5", children: [] },
        ],
      },
      { id: "6", children: [{ id: "7", children: [{ id: "8", children: [] }] }] },
    ];

    const tree2: TreeNode[] = [
      {
        id: "1",
        children: [
          { id: "2", children: [{ id: "9", children: [] }] },
          { id: "10", children: [] },
        ],
      },
      { id: "6", children: [{ id: "11", children: [{ id: "12", children: [] }] }] },
    ];

    const tree3: TreeNode[] = [{ id: "13", children: [{ id: "14", children: [] }] }];

    const mergedTrees = fn(tree1, tree2, tree3);

    expect(mergedTrees).toStrictEqual([
      {
        id: "1",
        children: [
          {
            id: "2",
            children: [
              { id: "3", children: [] },
              { id: "4", children: [] },
              { id: "9", children: [] },
            ],
          },
          { id: "5", children: [] },
          { id: "10", children: [] },
        ],
      },
      {
        id: "6",
        children: [
          { id: "7", children: [{ id: "8", children: [] }] },
          { id: "11", children: [{ id: "12", children: [] }] },
        ],
      },
      { id: "13", children: [{ id: "14", children: [] }] },
    ]);

    expect(mergedTrees).toMatchSnapshot();
  });
});
