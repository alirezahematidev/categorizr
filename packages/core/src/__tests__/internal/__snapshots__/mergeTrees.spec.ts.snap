// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mergeTrees > should returns the merged complex trees 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "id": "3",
          },
          {
            "children": [],
            "id": "4",
          },
          {
            "children": [],
            "id": "9",
          },
        ],
        "id": "2",
      },
      {
        "children": [],
        "id": "5",
      },
      {
        "children": [],
        "id": "10",
      },
    ],
    "id": "1",
  },
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "id": "8",
          },
        ],
        "id": "7",
      },
      {
        "children": [
          {
            "children": [],
            "id": "12",
          },
        ],
        "id": "11",
      },
    ],
    "id": "6",
  },
  {
    "children": [
      {
        "children": [],
        "id": "14",
      },
    ],
    "id": "13",
  },
]
`;

exports[`mergeTrees > should returns the merged trees 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "id": "2",
      },
      {
        "children": [],
        "id": "3",
      },
      {
        "children": [],
        "id": "6",
      },
    ],
    "id": "1",
  },
  {
    "children": [
      {
        "children": [],
        "id": "5",
      },
    ],
    "id": "4",
  },
  {
    "children": [
      {
        "children": [],
        "id": "8",
      },
    ],
    "id": "7",
  },
]
`;
